### PR TITLE
Rewrite of cy_variant.pyx for faster variant type conversion

### PR DESCRIPTION
**Pull request by hoytak**
This PR gives cy_variant.pyx the same love that cy_flexible_type.pyx
has received.

### CODE REVIEW

File: oss_src/unity/python/sframe/cython/cy_variant.pyx [hunk 1]
```diff
@@ -33,170 +36,593 @@ from .cy_unity cimport is_function_closure_info
...
+    # First, check if all the keys are string values.  If they are, we
+    # are doing a variant_map_type.  Otherwise, encode it as a
+    # flexible type.
+
+    cdef bint output_is_varmap = True
+
+    for k in d.iterkeys():
+        if type(k) is not str:
+            output_is_varmap = False
+            break
```
Comment (ylow) [inline oss_src/unity/python/sframe/cython/cy_variant.pyx]: I don't understand the logic here. If all keys are strings, it still can be a flexible_type via flex_dict.
The oddity is when the values contain non-trivial types like SArray/SFrame/etc.

Comment (hoytak) [inline oss_src/unity/python/sframe/cython/cy_variant.pyx]: Does flex_dict always have priority over varmap?   If that is the case, then I'll update this logic.

Comment (ylow) [inline oss_src/unity/python/sframe/cython/cy_variant.pyx]: May have to do something similar like the list case above: just repeated on keys() and values() possibly.

Comment (hoytak) [inline oss_src/unity/python/sframe/cython/cy_variant.pyx]: Can do.

Comment (ylow) [inline oss_src/unity/python/sframe/cython/cy_variant.pyx]: yeah. Generally, if we can contain something in a flexible_type, we should: it is the simpler type.

**Pull request merged**